File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int lmax = 0;
        int rmax = 0;
        int[] pre = new int[n];
        int[] post = new int[n];
        for(int i = 0; i < n; i++) {
            pre[i] = lmax;
            if(height[i] > lmax) lmax = height[i];
            post[n-i-1] = rmax;
            if(height[n-i-1] > rmax) rmax = height[n-i-1];
        }
        int totalWater = 0;
        for(int i = 0; i < height.length; i++) {
            if(Math.min(pre[i], post[i]) - height[i] > 0) {
                totalWater += Math.min(pre[i], post[i]) - height[i];
            }
        }
        return totalWater;
    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l, r = 0, n-1
        lmax = 0
        rmax = 0
        twater = 0
        while(l < r):
            lmax = max(lmax, height[l])
            rmax = max(rmax, height[r])
            if height[l] < height[r]:
                l += 1
                if lmax - height[l] > 0:
                    twater += lmax - height[l]
            else:
                r-=1
                if rmax - height[r] > 0:
                    twater += rmax - height[r]
        return twater